x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/ShrinkAction.java

```diff
@@ -85,31 +96,77 @@ public boolean isSafeAction() {
-            (index, clusterState) -> clusterState.getMetadata().index(index).getNumberOfShards() == numberOfShards);
+        BranchingStep conditionalSkipShrinkStep = new BranchingStep(preShrinkBranchingKey, waitForNoFollowerStepKey, nextStepKey,
+            getSkipShrinkStepPredicate(numberOfShards));
         WaitForNoFollowersStep waitForNoFollowersStep = new WaitForNoFollowersStep(waitForNoFollowerStepKey, readOnlyKey, client);
         UpdateSettingsStep readOnlyStep = new UpdateSettingsStep(readOnlyKey, setSingleNodeKey, client, readOnlySettings);
         SetSingleNodeAllocateStep setSingleNodeStep = new SetSingleNodeAllocateStep(setSingleNodeKey, allocationRoutedKey, client);
         CheckShrinkReadyStep checkShrinkReadyStep = new CheckShrinkReadyStep(allocationRoutedKey, shrinkKey);
         ShrinkStep shrink = new ShrinkStep(shrinkKey, enoughShardsKey, client, numberOfShards, SHRUNKEN_INDEX_PREFIX);
         ShrunkShardsAllocatedStep allocated = new ShrunkShardsAllocatedStep(enoughShardsKey, copyMetadataKey, SHRUNKEN_INDEX_PREFIX);
-        CopyExecutionStateStep copyMetadata = new CopyExecutionStateStep(copyMetadataKey, aliasKey, SHRUNKEN_INDEX_PREFIX,
-            ShrunkenIndexCheckStep.NAME);
+        CopyExecutionStateStep copyMetadata = new CopyExecutionStateStep(copyMetadataKey, dataStreamCheckBranchingKey,
+            SHRUNKEN_INDEX_PREFIX, ShrunkenIndexCheckStep.NAME);
+        // by the time we get to this step we have 2 indices, the source and the shrunken one. we now need to choose an index
+        // swapping strategy such that the shrunken index takes the place of the source index (which is also deleted).
+        // if the source index is part of a data stream it's a matter of replacing it with the shrunken index one in the data stream and
+        // then deleting the source index; otherwise we'll use the alias management api to atomically transfer the aliases from source to
+        // the shrunken index and delete the source
+        BranchingStep isDataStreamBranchingStep = new BranchingStep(dataStreamCheckBranchingKey, aliasKey, replaceDataStreamIndexKey,
+            (index, clusterState) -> {
+                IndexAbstraction indexAbstraction = clusterState.metadata().getIndicesLookup().get(index.getName());
+                assert indexAbstraction != null : "invalid cluster metadata. index [" + index.getName() + "] was not found";
+                return indexAbstraction.getParentDataStream() != null;
+            });
         ShrinkSetAliasStep aliasSwapAndDelete = new ShrinkSetAliasStep(aliasKey, isShrunkIndexKey, client, SHRUNKEN_INDEX_PREFIX);
+        ReplaceDataStreamBackingIndexStep replaceDataStreamBackingIndex = new ReplaceDataStreamBackingIndexStep(replaceDataStreamIndexKey,
+            deleteIndexKey, SHRUNKEN_INDEX_PREFIX);
+        DeleteStep deleteSourceIndexStep = new DeleteStep(deleteIndexKey, isShrunkIndexKey, client);
```

Comment:
I *think* (but I'm not sure), that this might fail for the alias case, because the index will already have been deleted in the `ShrinkSetAliasStep`, so we may need changes in `DeleteStep` to ignore deleting an index that doesn't exist

Reply:
on the non-data stream branch (ie. the alias handling) we don't try to delete the index using this step as the alias swap will delete it. This step is only used on the data stream handling branch, after the `replace-datastream-backing-index` step
```
        ShrinkSetAliasStep aliasSwapAndDelete = new ShrinkSetAliasStep(aliasKey, isShrunkIndexKey, client, SHRUNKEN_INDEX_PREFIX);
```
The `isShrunkIndexKey` step is next on the alias handling branch.